### ISSUE TITLE
[mongodb-9.2.4] replicaset initialization fails with externalAccess Nodeport and Domain

### ISSUE THREAD

**Issue by horihel** [bitnami/charts#4073]
**Which chart**:
mongodb-9.2.4

**Describe the bug**
I still don't understand exactly - it looks like the chart sets MONGODB_ADVERTISED_HOSTNAME to externalAccess.service.domain and the primary config script tries to use that to construct the replicaSet. But even if the external address is available from inside the pod, the port numbers are not communicated to mongo, so it can't reach it.

Maybe the setup script should not use ADVERTISED_HOSTNAME on initial setup?

BTW: setting externalAccess.enabled: false, deploying it and _then_ change to true works fine.

**To Reproduce**
Install chart with these values:
```yaml
auth:
  enabled: false
architecture: replicaset
image:
  debug: true
externalAccess:
  enabled: true
  service:
    type: NodePort
    domain: '192.168.172.132'
    nodePorts:
    - 30001
    - 30002
```

MongoDB will not initialize correctly:
```log
Advertised Hostname: 192.168.172.132
Pod name matches initial primary pod name, configuring node as a primary
mongodb 13:36:23.10
mongodb 13:36:23.10 Welcome to the Bitnami mongodb container
mongodb 13:36:23.10 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-mongodb
mongodb 13:36:23.10 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-mongodb/issues
mongodb 13:36:23.10
mongodb 13:36:23.10 INFO ==> ** Starting MongoDB setup **
mongodb 13:36:23.11 INFO ==> Validating settings in MONGODB_* env vars...
mongodb 13:36:23.11 WARN ==> You set the environment variable ALLOW_EMPTY_PASSWORD=yes. For safety reasons, do not use this flag in a production environment.
mongodb 13:36:23.12 INFO ==> Initializing MongoDB...
mongodb 13:36:23.14 INFO ==> Deploying MongoDB from scratch...
mongodb 13:36:23.14 DEBUG ==> Starting MongoDB in background...
about to fork child process, waiting until server is ready for connections.
forked process: 41
child process started successfully, parent exiting
mongodb 13:36:23.92 INFO ==> Creating users...
mongodb 13:36:23.92 INFO ==> Users created
mongodb 13:36:23.94 INFO ==> Configuring MongoDB replica set...
mongodb 13:36:23.94 INFO ==> Stopping MongoDB...
mongodb 13:36:24.95 DEBUG ==> Starting MongoDB in background...
about to fork child process, waiting until server is ready for connections.
forked process: 116
child process started successfully, parent exiting
mongodb 13:36:25.90 INFO ==> Configuring MongoDB primary node
mongodb 13:39:25.58 ERROR ==> MongoDB primary node failed to get configured
mongodb 13:39:25.59 INFO ==> Stopping MongoDB...
```

Pods report healthy after restart, but the replicaset is busted. rs.status() reports:
```json
{
        "operationTime" : Timestamp(0, 0),
        "ok" : 0,
        "errmsg" : "no replset config has been received",
        "code" : 94,
        "codeName" : "NotYetInitialized",
        "$clusterTime" : {
                "clusterTime" : Timestamp(0, 0),
                "signature" : {
                        "hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
                        "keyId" : NumberLong(0)
                }
        }
}
```

**Expected behavior**

Recplicaset gets constructed correctly and mongodb is ready to work.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"clean", GoVersion:"go1.14.9"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-20T14:12:21Z", GoVersion:"go1.15.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:07:13Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**

**Comment by horihel** [comment]
further testing revealed that enabling auth also works, but the domain setting does not seem to have any influence.
so auth.enabled: false and externalAccess.enabled:true -> fail to init.

**Comment by joancafom** [comment]
Hi @horihel 

Thanks for your issue! I have tried to replicate it using the values you provided before and did not run into any problems (I'm using Minikube). What cluster are you using?

**Comment by horihel** [comment]
we're using rancher (2.3) as a single-node configuration (this is meant to be a developer-VM - that's why we're running this without auth)

**Comment by dani8art** [comment]
Hi @horihel after some investigation a run into the issue and I was able to reproduce it, ass you said 

It seems we are adding the wrong nodes' ports to the replicaset when using external access

https://github.com/bitnami/bitnami-docker-mongodb/blob/de4c58a2eed72cb9fddce5ab631ce8a12a097e30/4.4/debian-10/rootfs/opt/bitnami/scripts/libmongodb.sh#L566

https://github.com/bitnami/bitnami-docker-mongodb/blob/de4c58a2eed72cb9fddce5ab631ce8a12a097e30/4.4/debian-10/rootfs/opt/bitnami/scripts/libmongodb.sh#L594

https://github.com/bitnami/bitnami-docker-mongodb/blob/de4c58a2eed72cb9fddce5ab631ce8a12a097e30/4.4/debian-10/rootfs/opt/bitnami/scripts/libmongodb.sh#L621

I think the solution would be to either use the internal access to create the replicaset and then use external access as entry point or try to set the ports properly by adding an env like `MONGODB_ADVERTISED_PORT`. I have doubts about how the first approach would work since the replicaset is configured using internal IPs but after that, we will use it through the external one. This is something that we would need to tests. 

BTW, we could the workaround you mentioned before, it worked perfectly for me (but it has the same problem as the first approach)

Thank you so much for your feedback. We have opened an internal task in order to evaluate it, unfortunately, since there could be a way to make it work and regarding priorities and internal milestones we can't give you an ETA. However, if it is important for you and you have a chance to create a PR we would be happy to handle it.

**Comment by horihel** [comment]
my gut feeling is that the chart should not assume that the address(es) in externalAccess are reachable by the pod at all. The cluster (namespace) might not be allowed to connect to external addresses itself (but might be still reachable on those addresses).

libmongodb.sh and the bitnami scripts i probably need to understand first to dare change something. I'll try to allocate time for it. Thanks for pointing out the place to look for!

**Comment by iosifnicolae2** [comment]
I've managed to fix this problem by updating the `setup.sh` script as follow:
https://github.com/bitnami/charts/blob/5f4b4b1f9aec59371a16ead48f01c81b5cdfb09f/bitnami/mongodb/templates/replicaset/scripts-configmap.yaml#L54
to:
```bash
#!/bin/bash
. /opt/bitnami/scripts/libnet.sh
export MONGODB_ADVERTISED_HOSTNAME=$(get_mongo_hostname)

echo "Advertised Hostname: $MONGODB_ADVERTISED_HOSTNAME"

if [[ "$MY_POD_NAME" = "mongodb-0" ]]; then
echo "Pod name matches initial primary pod name, configuring node as a primary"
export MONGODB_REPLICA_SET_MODE="primary"
else
echo "Pod name doesn't match initial primary pod name, configuring node as a secondary"
export MONGODB_REPLICA_SET_MODE="secondary"
export MONGODB_INITIAL_PRIMARY_ROOT_PASSWORD="$MONGODB_ROOT_PASSWORD"
export MONGODB_INITIAL_PRIMARY_PORT_NUMBER="$MONGODB_PORT_NUMBER"
export MONGODB_ROOT_PASSWORD="" MONGODB_USERNAME="" MONGODB_DATABASE="" MONGODB_PASSWORD=""
fi

exec /opt/bitnami/scripts/mongodb/entrypoint.sh /opt/bitnami/scripts/mongodb/run.sh
```

Thanks @dani8art for your suggestions!

**Comment by horihel** [comment]
Hello @iosifnicolae2 - (regarding to PS2) excuse my question, but for this use case, the containers actually _are_ exposed to the public (via using externalAccess) - or did I misunderstand?

I'm just afraid any change may break ADVERTISED_HOSTNAMES with externalAccess on real production installations, because when using mongodb://[...]/?replicaSet=rs0 it's important that mongo returns valid reachable addresses to clients (internal or external)

**Comment by iosifnicolae2** [comment]
> Hello @iosifnicolae2 - (regarding to PS2) excuse my question, but for this use case, the containers actually _are_ exposed to the public (via using externalAccess) - or did I misunderstand?
> 
> I'm just afraid any change may break ADVERTISED_HOSTNAMES with externalAccess on real production installations, because when using mongodb://[...]/?replicaSet=rs0 it's important that mongo returns valid reachable addresses to clients (internal or external)

Yes, you're right.

I have a different use-case (I need my MongoDB instance to be exposed just on the internal network..).

**Comment by marcosbc** [comment]
@horihel Thanks for the update, I've added your notes to the internal task. It would be awesome if you could find the time to look into this, as we currently have a very limited bandwidth (due to a huge backlog and winter vacations). We'd be happy to review any related PRs!

**Comment by alemorcuq** [comment]
Hi, @horihel.

I'm looking into this and I have a question. Would it work correctly if you configure the replica set using internal IPs and then create the NodePort services to access them externally? I mean, all the configuration would be done with internal IPs but you would be able to connect externally via those NodePort services. I believe this is something similar to what you said about installing the chart with external access disabled and then enabling it.

I'm thinking about these changes:

```diff --git a/bitnami/mongodb/templates/replicaset/scripts-configmap.yaml b/bitnami/mongodb/templates/replicaset/scripts-configmap.yaml
index f6b197cd7..119c5da28 100644
--- a/bitnami/mongodb/templates/replicaset/scripts-configmap.yaml
+++ b/bitnami/mongodb/templates/replicaset/scripts-configmap.yaml
@@ -62,12 +62,6 @@ data:
     ID="${MY_POD_NAME#"{{ $fullname }}-"}"
     export MONGODB_ADVERTISED_HOSTNAME=$(echo '{{ .Values.externalAccess.service.loadBalancerIPs }}' | tr -d '[]' | cut -d ' ' -f "$(($ID + 1))")
     {{- end }}
-    {{- else if eq .Values.externalAccess.service.type "NodePort" }}
-    {{- if .Values.externalAccess.service.domain }}
-    export MONGODB_ADVERTISED_HOSTNAME={{ .Values.externalAccess.service.domain }}
-    {{- else }}
-    export MONGODB_ADVERTISED_HOSTNAME=$(curl -s https://ipinfo.io/ip)
-    {{- end }}
     {{- end }}
     {{- end }}

diff --git a/bitnami/mongodb/templates/replicaset/statefulset.yaml b/bitnami/mongodb/templates/replicaset/statefulset.yaml
index ee793bd52..3f3d58558 100644
--- a/bitnami/mongodb/templates/replicaset/statefulset.yaml
+++ b/bitnami/mongodb/templates/replicaset/statefulset.yaml
@@ -207,7 +207,7 @@ spec:
               value: "{{ include "mongodb.fullname" . }}-0.$(K8S_SERVICE_NAME).$(MY_POD_NAMESPACE).svc.{{ .Values.clusterDomain }}"
             - name: MONGODB_REPLICA_SET_NAME
               value: {{ .Values.replicaSetName | quote }}
-            {{- if and .Values.replicaSetHostnames (not .Values.externalAccess.enabled) }}
+            {{- if and .Values.replicaSetHostnames (or (not .Values.externalAccess.enabled) (eq .Values.externalAccess.service.type "NodePort")) }}
             - name: MONGODB_ADVERTISED_HOSTNAME
               value: "$(MY_POD_NAME).$(K8S_SERVICE_NAME).$(MY_POD_NAMESPACE).svc.{{ .Values.clusterDomain }}"
             {{- end }}```

**Comment by horihel** [comment]
This might do the trick.
Does this remove the ipinfo detection logic only from initial setup or completely?

**Comment by alemorcuq** [comment]
> This might do the trick.
> Does this remove the ipinfo detection logic only from initial setup or completely?

Completely, I'm afraid. Am I missing something?

**Comment by horihel** [comment]
i read a little bit of mongo docs and the mongo connection logic seems to enforce that clients _and_ replicaset members must be able to reach themselves with the same (preferredly DNS) address.

i guess with that in mind (and if we want to keep connecting with seed (?replicaset=rs0) working) mongo actually _must_ initiate the cluster with the external adress(es) and the admin must make sure that the external address is reachable by all pods.

Because AFAICS mongo doesn't know a difference between replicaset member addresses and "advertised" adresses. (and i hope i'm wrong here).

The only other solution i can think of is hotpatching /etc/hosts in the nodes every time a pod changes ClusterIP - which is probably quite a bit of work.

So in conclusion, the correct solution is all the way up in this thread from @dani8art : the setup script needs to know the correct port number to allow mongo to connect to itself.

**Comment by juan131** [comment]
> i guess with that in mind (and if we want to keep connecting with seed (?replicaset=rs0) working) mongo actually must initiate the cluster with the external address(es) and the admin must make sure that the external address is reachable by all pods.

+1 

I as think the solutions consists on configuring MongoDB with the external addresses to avoid it advertising the internal ones.

**Comment by HassanHashemi** [comment]
Any stable workaround for this?

**Comment by idonca** [comment]
what we did: create another service with NodePort type:

```yaml
{{- if eq .Values.architecture "replicaset" }}
apiVersion: v1
kind: Service
metadata:
  name: "{{ include "mongodb.service.nameOverride" . }}-nodeport"
  namespace: {{ include "mongodb.namespace" . }}
  labels: {{- include "common.labels.standard" . | nindent 4 }}
    app.kubernetes.io/component: mongodb
  {{- if .Values.service.annotations }}
  annotations: {{- include "common.tplvalues.render" (dict "value" .Values.service.annotations "context" $) | nindent 4 }}
  {{- end }}
spec:
  type: NodePort
  ports:
    - name: {{ .Values.service.portName }}
      port: {{ .Values.service.port }}
      targetPort: mongodb
      nodePort: 30007
  selector: {{- include "common.labels.matchLabels" . | nindent 4 }}
    app.kubernetes.io/component: mongodb
{{- end }}
```
And with this, you will have two services, one that has the scope to be internal(ClusterIP) and the other one to be external(NodePort).

**Comment by HassanHashemi** [comment]
@idonca it's not possible to connect to replica set. though it's possible to connect as "Direct Connection".

**Comment by pixelfields** [comment]
The issue is opened more than 1 year. Any plans to work on it?

We hit this recently, tried to deploy to EKS with LoadBalancer external access configured. Our app is built on the top of SpringBoot, so using the direct access is not a valid solution for us at the moment.

**Comment by alemorcuq** [comment]
We have a task to tackle this, but there's no ETA. I will close this thread as there's a newer one on this issue here #5508.
